testutil/integration/longwait_dkg_test.go

```diff
@@ -0,0 +1,184 @@
+// | Node 0 | ^^^^^^^^^^ | ^^^^^^^^^^ | ^^^^^^^^^^ | ^^^^^^^^^^ |
+// +--------+------------+------------+------------+------------+
+// | Node 1 | ---------- | ^^^--^^^^^ | ^^^^^^--^^ | ^^^^^^^^^^ |
+// +--------+------------+------------+------------+------------+
+// | Node 2 | ---------- | ---------- | ^--^^^^^^^ | ^^^^^^^^^^ |
+// +--------+------------+------------+------------+------------+
+// | Node 3 | ---------- | ---------- | ---------- | ^^^^^^^^^^ |
+// +--------+------------+------------+------------+------------+
+
+func TestLongWaitDKG(t *testing.T) {
+	if !*longwaitDKG {
+		t.Skip("Long wait test is disabled")
+	}
+
+	const (
+		threshold = 3
+		numNodes  = 4
+		numVals   = 10
+		relayURL  = "https://0.relay.obol.tech"
+		// The time period after which a new node is started.
+		window = 10 * time.Second
+		// The time period when a node is offline. A window is divided into multiple slots each of nodeDownPeriod duration.
+		// A node goes down in one of these slots in a particular window.
+		nodeDownPeriod = 2 * time.Second
+	)
+
+	var (
+		eg              errgroup.Group
+		ctx             = log.WithTopic(context.Background(), "longwait")
+		allNodesStarted = make(chan struct{}, numNodes)
+	)
+
+	def, p2pKeys := testDef(t, threshold, numNodes, numVals)
+	dir := t.TempDir()
+
+	for i, p2pKey := range p2pKeys {
+		indx := i
+		p2pKey := p2pKey
+		noRestart := indx == 0
+		if indx > 0 {
+			time.Sleep(window)
+		}
+
+		eg.Go(func() error {
+			nodeDir := path.Join(dir, fmt.Sprintf("node%d", indx))
+			require.NoError(t, os.Mkdir(nodeDir, 0o750))
+
+			dkgConf := dkg.Config{
+				DataDir: nodeDir,
+				P2P: p2p.Config{
+					Relays: []string{relayURL},
+				},
+				Log:           log.DefaultConfig(),
+				ShutdownDelay: 1 * time.Second,
+				TestConfig: dkg.TestConfig{
+					Def:    &def,
+					P2PKey: p2pKey,
+				},
+			}
+
+			log.Debug(ctx, "Starting node (1st time)", z.Int("node", indx))
+
+			return mimicDKGNode(ctx, t, noRestart, dkgConf, window, nodeDownPeriod, indx, allNodesStarted)
+		})
+	}
+
+	// Notify all nodes that everyone has started. Let the DKG begin!
+	for i := 0; i < numNodes; i++ {
+		allNodesStarted <- struct{}{}
+	}
+
+	require.NoError(t, eg.Wait())
+}
+
+// mimicDKGNode mimics the behaviour of a DKG node that randomly stops for sometime before restarting again but finally participates in the DKG.
+func mimicDKGNode(ctx context.Context, t *testing.T, noRestart bool, dkgConf dkg.Config, window, nodeDownPeriod time.Duration, nodeIdx int, allNodesStarted chan struct{}) error {
+	t.Helper()
+
+	if noRestart {
+		return dkg.Run(ctx, dkgConf)
+	}
+
+	for {
+		var allStarted bool
+		select {
+		case <-allNodesStarted:
+			allStarted = true
+		default:
+			break
+		}
+
+		if allStarted { // Do the final DKG since all nodes are up now
+			break
+		}
+
+		ctx, cancel := context.WithCancel(ctx)
+		go func(ctx context.Context) {
+			_ = dkg.Run(ctx, dkgConf)
```

Comment:
Just as a safety measure, I would advice to check this error and require that it originates from `ctx` being cancelled.

On top of that to avoid any kind of conflict, maybe after calling `cancel()` wait that this goroutine effectively finished?